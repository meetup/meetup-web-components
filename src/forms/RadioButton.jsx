import React from 'react';
import PropTypes from 'prop-types';
import { Radio as SwarmRadio } from '@meetup/swarm-components';

const RadioButton = ({ label, name, value, ...other }) => (
	<SwarmRadio
		name={name}
		value={value}
		label={label}
		id={`RadioButton-${name}-${value}`}
		{...other}
	/>
);

RadioButton.displayName = 'RadioButton';

RadioButton.propTypes = {
	/** What we render into the input's `<label />` */
	label: PropTypes.string.isRequired,

	/** Additional class name/s to add to the wrapper element  */
	className: PropTypes.string,
	/**
	 * Used to associate a group of radio buttons.
	 * Only one radio button in a group can be selected.
	 */
	name: PropTypes.string,
	/**
	 * Value of the input.
	 */
	value: PropTypes.string,
};

export default RadioButton;
