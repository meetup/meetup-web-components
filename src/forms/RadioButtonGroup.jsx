import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

/**
 * @class RadioButtonGroup
 */
export default class RadioButtonGroup extends PureComponent {
	constructor(props) {
		super(props);

		const initialState = {
			value: props.selectedValue,
		};

		React.Children.forEach(props.children, radio => {
			initialState[radio.props.value] = radio.props.value === props.selectedValue;
		});

		this.state = initialState;
		this.onChange = this.onChange.bind(this);
		this.isSelected = this.isSelected.bind(this);
	}

	componentWillReceiveProps(nextProps) {
		if ('selectedValue' in nextProps) {
			const newState = {};

			Object.keys(this.state).forEach(stateKey => {
				if (stateKey === 'value') {
					newState.value = nextProps.selectedValue;
				} else {
					newState[stateKey] = stateKey === nextProps.selectedValue;
				}
			});

			this.setState(newState);
		}
	}

	onChange(e) {
		this.props.onChange && this.props.onChange(e);

		// TODO: also update state of individual buttons
		this.setState({
			value: e.target.value,
		});
	}

	isSelected(radio) {
		return radio.props.value === this.state.value;
	}

	render() {
		const { children, className } = this.props;

		return (
			<div className={className}>
				{React.Children.map(children, radio =>
					React.cloneElement(radio, {
						onChange: this.onChange,
						checked: this.isSelected(radio),
						...radio.props,
					})
				)}
			</div>
		);
	}
}

RadioButtonGroup.propTypes = {
	children: PropTypes.arrayOf(PropTypes.element).isRequired,

	/** The value of the selected input in the radio group */
	selectedValue: PropTypes.string,

	/** The name attribute on each radio in the group */
	name: PropTypes.string.isRequired,

	/** Additional class name/s to add to the wrapper element */
	className: PropTypes.string,

	/** Callback that happens when the input changes */
	onChange: PropTypes.func,
};
