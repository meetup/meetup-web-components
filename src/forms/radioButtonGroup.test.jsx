import React from 'react';
import { shallow } from 'enzyme';

import RadioButtonGroup from './RadioButtonGroup';
import RadioButton from './RadioButton';

const DEFAULT_PROPS = {
	className: 'some-class',
	selectedValue: 'one',
	name: 'name',
};

const getWrapper = (props = {}) =>
	shallow(
		<RadioButtonGroup {...DEFAULT_PROPS} {...props}>
			<RadioButton value="one" label="label 1" name="button1" />
			<RadioButton value="two" label="label 2" name="button2" />
		</RadioButtonGroup>
	);

describe('RadioButtonGroup', () => {
	it('should match the snapshot', () => {
		expect(getWrapper()).toMatchSnapshot();
	});

	describe('componentWillReceiveProps', () => {
		it('should update state.value if the next props contains selectedValue', () => {
			const wrapper = getWrapper({
				selectedValue: 'one',
			});
			const instance = wrapper.instance();

			instance.componentWillReceiveProps({ selectedValue: 'two' });
			expect(wrapper.state('selectedValue')).toBe('two');
		});

		it('should not update state for nextProps other than selectedValue', () => {
			const wrapper = getWrapper();
			const instance = wrapper.instance();

			jest.spyOn(instance, 'setState');
			instance.componentWillReceiveProps({ testProp: true, other: 'other' });
			expect(instance.setState).not.toHaveBeenCalled();
		});
	});

	describe('handleChange', () => {
		const MOCK_EVENT = {
			target: {
				value: 'two',
			},
		};

		it('should update the state for selectedValue', () => {
			const wrapper = getWrapper({
				selectedValue: 'one',
			});
			const instance = wrapper.instance();
			instance.handleChange(MOCK_EVENT);
			expect(wrapper.state('selectedValue')).toBe('two');
		});

		it('should call the onChange prop', () => {
			const onChange = jest.fn();
			const wrapper = getWrapper({ onChange });
			const instance = wrapper.instance();
			instance.handleChange(MOCK_EVENT);
			expect(onChange).toHaveBeenCalledWith(MOCK_EVENT);
		});
	});
});
