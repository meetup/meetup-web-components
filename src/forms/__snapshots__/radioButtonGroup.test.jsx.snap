// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtonGroup should match the snapshot 1`] = `
<div
  className="some-class"
>
  <RadioButton
    checked={true}
    key=".0"
    label="label 1"
    name="button1"
    onChange={[Function]}
    value="one"
  />
  <RadioButton
    checked={false}
    key=".1"
    label="label 2"
    name="button2"
    onChange={[Function]}
    value="two"
  />
</div>
`;
